packages/ckb-indexer/src/ckbIndexerFilter.ts

```diff
@@ -0,0 +1,341 @@
+  }
+
+  if (queryOptions.fromBlock || queryOptions.toBlock) {
+    searchKey.filter!.blockRange = [
+      queryOptions.fromBlock || "0x0",
+      queryOptions.toBlock
+        ? BI.from(queryOptions.toBlock).add(1).toHexString()
+        : "0x" + Number.MAX_SAFE_INTEGER.toString(16),
+    ];
+  }
+
+  return searchKey;
+}
+
+function filterByQueryOptions(cells: Cell[], options: QueryOptions): Cell[] {
+  const searchKey = convertQueryOptionToSearchKey(options);
+  let filteredCells = cells.filter((cell) => filterBy(cell, searchKey));
+
+  if (options.argsLen && options.argsLen !== "any") {
+    filteredCells = filteredCells.filter(
+      (cell) =>
+        bytes.bytify(cell.cellOutput.lock.args).length === options.argsLen
+    );
+  }
+
+  if (!!options.data && options.data !== "any") {
+    if (
+      instanceOfDataWithSearchMode(options.data) &&
+      options.data.searchMode === "exact"
+    ) {
+      const dataSearch = options.data as DataWithSearchMode;
+      filteredCells = filteredCells.filter((cell) =>
+        bytes.equal(bytes.bytify(cell.data), bytes.bytify(dataSearch.data))
+      );
+    } else if (
+      instanceOfDataWithSearchMode(options.data) &&
+      options.data.searchMode === "prefix"
+    ) {
+      const dataSearch = options.data as DataWithSearchMode;
+      filteredCells = filteredCells.filter(
+        (cell) =>
+          Buffer.from(bytes.bytify(cell.data)).indexOf(
+            bytes.bytify(dataSearch.data)
+          ) === 0
+      );
+    } else {
+      filteredCells = filteredCells.filter(
+        (cell) =>
+          Buffer.from(bytes.bytify(cell.data)).indexOf(
+            bytes.bytify(options.data as string)
+          ) === 0
+      );
+    }
+  }
+
+  if (options.order === "desc") {
+    filteredCells.reverse();
+  }
+
+  if (options.skip) {
+    filteredCells = filteredCells.slice(options.skip);
+  }
+
+  return filteredCells;
+}
+
+function filterBy(cell: Cell, searchKey: SearchKey): boolean {
+  const isExactMode = searchKey.scriptSearchMode === "exact";
+  const { cellOutput } = cell;
+  const { scriptType, script, filter } = searchKey;
+
+  // Search mode
+  if (isExactMode) {
+    if (scriptType === "lock") {
+      if (
+        !bytes.equal(
+          blockchain.Script.pack(cellOutput.lock),
+          blockchain.Script.pack(script)
+        )
+      ) {
+        return false;
+      }
+    } else {
+      if (
+        !cellOutput.type ||
+        !bytes.equal(
+          blockchain.Script.pack(cellOutput.type),
+          blockchain.Script.pack(script)
+        )
+      ) {
+        return false;
+      }
+    }
+    // Prefix mode
+  } else {
+    if (scriptType === "lock") {
+      if (!checkScriptWithPrefixMode(cellOutput.lock, script)) {
+        return false;
+      }
+    } else {
+      if (!checkScriptWithPrefixMode(cellOutput.type, script)) {
+        return false;
+      }
+    }
+  }
+
+  // Filter is always in prefix mode
+  if (filter) {
```

Comment:
```ts
if(!filter) return true;
```

early return makes the code more readable

Reply:
updated